### ISSUE TITLE
Handle server errors

### ISSUE THREAD

**Issue by dos1** [astrada/google-drive-ocamlfuse#147]
Hello,

I've sent pull request #146 to fix an issue with GDrive rejecting tar.gz files when sent without proper mime type, thanks for fast merge!

That bug has bitten us quite painfully, since files were just being copied as usual with no errors being reported. We noticed that something is wrong when we tried to read the files and it turned out that all of the files stored after Nov 20 has been simply empty. Aside of that, absolutely nothing suggested that there's any error happening anywhere.

After turning on curl logs, it turned out that Google Drive server has been returning 403 Forbidden with following error:
```
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "forbidden",
    "message": "The requested mime type change is forbidden.",
    "locationType": "other",
    "location": "file.mimeType"
   }
  ],
  "code": 403,
  "message": "The requested mime type change is forbidden."
 }
}
```

Would it be possible to return some filesystem error when the transfer has so obviously failed? Or, it it's done asynchronously, could failed files be at least stored somewhere and logged as error even without debug mode, so they're not silently lost?

**Comment by zardii** [comment]
Being in no way whatsoever affiliated with the previous developer (other than working in the same company) I have to say that the bug was very painful for us indeed ;]

**Comment by astrada** [comment]
I will do a couple of tests to check why the error was ignored. Thanks for reporting it.

**Comment by astrada** [comment]
I'm sorry but I was not able to reproduce the issue. When I uploaded a .tar.gz with application/octet-stream as MIME type, the file was uploaded correctly and the type was corrected by Drive to application/x-gzip. For example:

1) POST (file creation):

    [696.412495] curl: header out: POST /drive/v2/files?fields=alternateLink%2CcreatedDate%2CdownloadUrl%2Ceditable%2Cetag%2CexplicitlyTrashed%2CexportLinks%2CfileExtension%2CfileSize%2Cid%2Clabels%2ClastViewedByMeDate%2Cmd5Checksum%2CmimeType%2CmodifiedDate%2Cparents%2Ctitle HTTP/1.1
    User-Agent: google-drive-ocamlfuse (0.5.19) gapi-ocaml/0.2.6/Unix
    Host: www.googleapis.com
    Accept: */*
    Accept-Encoding: identity
    Content-Type: application/json
    Authorization: Bearer 
    Content-Length: 120

    [696.412600] curl: data out: {"mimeType":"application/octet-stream","parents":[{"id":"0B-TenPiSMBplblFKWHQxZGpLaDA"}],"title":"polymap-0.1-2.tar.gz"}
    [697.122451] curl: header in: HTTP/1.1 200 OK
    [697.122540] curl: header in: Cache-Control: no-cache, no-store, max-age=0, must-revalidate
    [697.122621] curl: header in: Pragma: no-cache
    [697.122637] curl: header in: Expires: Fri, 01 Jan 1990 00:00:00 GMT
    [697.122650] curl: header in: Date: Mon, 30 Nov 2015 21:37:42 GMT
    [697.122663] curl: header in: Vary: X-Origin
    [697.122677] curl: header in: Content-Type: application/json; charset=UTF-8
    [697.122689] curl: header in: X-Content-Type-Options: nosniff
    [697.122698] curl: header in: X-Frame-Options: SAMEORIGIN
    [697.122709] curl: header in: X-XSS-Protection: 1; mode=block
    [697.122723] curl: info: Server GSE is not blacklisted
    [697.122732] curl: header in: Server: GSE
    [697.122742] curl: header in: Alternate-Protocol: 443:quic,p=1
    [697.122752] curl: header in: Alt-Svc: quic=":443"; ma=604800; v="30,29,28,27,26,25"
    [697.122763] curl: header in: Accept-Ranges: none
    [697.122792] curl: header in: Vary: Origin,Accept-Encoding
    [697.122805] curl: header in: Transfer-Encoding: chunked
    [697.122816] curl: header in: 
    [697.122829] curl: data in: 4f1
    {
     "id": "0B-TenPiSMBplUTNya2VoVkI5cGc",
     "etag": "\"fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA\"",
     "alternateLink": "https://drive.google.com/file/d/0B-TenPiSMBplUTNya2VoVkI5cGc/view?usp=drivesdk",
     "title": "polymap-0.1-2.tar.gz",
     "mimeType": "application/octet-stream",
     "labels": {
      "starred": false,
      "hidden": false,
      "trashed": false,
      "restricted": false,
      "viewed": true
     },
     "createdDate": "2015-11-30T21:37:41.634Z",
     "modifiedDate": "2015-11-30T21:37:41.634Z",
     "lastViewedByMeDate": "2015-11-30T21:37:41.634Z",
     "parents": [
      {
       "kind": "drive#parentReference",
       "id": "0B-TenPiSMBplblFKWHQxZGpLaDA",
       "selfLink": "https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc/parents/0B-TenPiSMBplblFKWHQxZGpLaDA",
       "parentLink": "https://www.googleapis.com/drive/v2/files/0
    [697.122902] curl: data in: B-TenPiSMBplblFKWHQxZGpLaDA",
       "isRoot": false
      }
     ],
     "downloadUrl": "https://doc-0k-a0-docs.googleusercontent.com/docs/securesc/2jstq75ksd6djnng7gd4lrkuiaklkreg/ha6129fgd7ptruqhv9htgq5jgcfrpv65/1448913600000/01371704680682440553/01371704680682440553/0B-TenPiSMBplUTNya2VoVkI5cGc?e=download&gd=true",
     "fileExtension": "gz",
     "md5Checksum": "d41d8cd98f00b204e9800998ecf8427e",
     "fileSize": "0",
     "editable": true,
     "explicitlyTrashed": false
    }

2) GET:

    [699.032208] curl: header out: GET /drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc?fields=alternateLink%2CcreatedDate%2CdownloadUrl%2Ceditable%2Cetag%2CexplicitlyTrashed%2CexportLinks%2CfileExtension%2CfileSize%2Cid%2Clabels%2ClastViewedByMeDate%2Cmd5Checksum%2CmimeType%2CmodifiedDate%2Cparents%2Ctitle HTTP/1.1
    User-Agent: google-drive-ocamlfuse (0.5.19) gapi-ocaml/0.2.6/Unix
    Host: www.googleapis.com
    Accept: */*
    Accept-Encoding: identity
    Authorization: Bearer 

    [699.298750] curl: header in: HTTP/1.1 200 OK
    [699.298816] curl: header in: Expires: Mon, 30 Nov 2015 21:37:44 GMT
    [699.298829] curl: header in: Date: Mon, 30 Nov 2015 21:37:44 GMT
    [699.298840] curl: header in: Cache-Control: private, max-age=0, must-revalidate, no-transform
    [699.298850] curl: header in: ETag: "fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA"
    [699.298860] curl: header in: Vary: Origin
    [699.298871] curl: header in: Vary: X-Origin
    [699.298883] curl: header in: Content-Type: application/json; charset=UTF-8
    [699.298894] curl: header in: X-Content-Type-Options: nosniff
    [699.298905] curl: header in: X-Frame-Options: SAMEORIGIN
    [699.298915] curl: header in: X-XSS-Protection: 1; mode=block
    [699.298927] curl: header in: Content-Length: 1265
    [699.298942] curl: info: Server GSE is not blacklisted
    [699.298953] curl: header in: Server: GSE
    [699.298964] curl: header in: Alternate-Protocol: 443:quic,p=1
    [699.299070] curl: header in: Alt-Svc: quic=":443"; ma=604800; v="30,29,28,27,26,25"
    [699.299107] curl: header in: 
    [699.299884] curl: data in: {
     "id": "0B-TenPiSMBplUTNya2VoVkI5cGc",
     "etag": "\"fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA\"",
     "alternateLink": "https://drive.google.com/file/d/0B-TenPiSMBplUTNya2VoVkI5cGc/view?usp=drivesdk",
     "title": "polymap-0.1-2.tar.gz",
     "mimeType": "application/octet-stream",
     "labels": {
      "starred": false,
      "hidden": false,
      "trashed": false,
      "restricted": false,
      "viewed": true
     },
     "createdDate": "2015-11-30T21:37:41.634Z",
     "modifiedDate": "2015-11-30T21:37:41.634Z",
     "lastViewedByMeDate": "2015-11-30T21:37:41.634Z",
     "parents": [
      {
       "kind": "drive#parentReference",
       "id": "0B-TenPiSMBplblFKWHQxZGpLaDA",
       "selfLink": "https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc/parents/0B-TenPiSMBplblFKWHQxZGpLaDA",
       "parentLink": "https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplblFKWHQxZGpLaDA",
       "isRoot": false
      }
     ],
     "downloadUrl": "https://doc-0k-a0-docs.googleusercontent.com/docs/securesc/2jstq75ksd6djnng7gd4lrkuiaklkreg/ha6129fgd7ptruqhv9htgq5jgcfrpv65/1448913600000/01371704680682440553/01371704680682440553/0B-TenPiSMBplUTNya2VoVkI5cGc?e=download&gd=true",
     "fileExtension": "gz",
     "md5Checksum": "d41d8cd98f00b204e9800998ecf8427e",
     "fileSize": "0",
     "editable": true,
     "explicitlyTrashed": false
    }

3) PUT:

    [699.300700] curl: header out: PUT /resumable/upload/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc?fields=alternateLink%2CcreatedDate%2CdownloadUrl%2Ceditable%2Cetag%2CexplicitlyTrashed%2CexportLinks%2CfileExtension%2CfileSize%2Cid%2Clabels%2ClastViewedByMeDate%2Cmd5Checksum%2CmimeType%2CmodifiedDate%2Cparents%2Ctitle&updateViewedDate=true HTTP/1.1
    User-Agent: google-drive-ocamlfuse (0.5.19) gapi-ocaml/0.2.6/Unix
    Host: www.googleapis.com
    Accept: */*
    Accept-Encoding: identity
    Transfer-Encoding: chunked
    Content-Type: application/json
    Authorization: Bearer 
    If-Match: "fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA"
    X-Upload-Content-Type: application/octet-stream
    X-Upload-Content-Length: 1997
    Expect: 100-continue

    [699.367056] curl: header in: HTTP/1.1 100 Continue
    [699.367291] curl: data out: 40e
    {"alternateLink":"https://drive.google.com/file/d/0B-TenPiSMBplUTNya2VoVkI5cGc/view?usp=drivesdk","createdDate":"2015-11-30T21:37:41.000Z","downloadUrl":"https://doc-0k-a0-docs.googleusercontent.com/docs/securesc/2jstq75ksd6djnng7gd4lrkuiaklkreg/ha6129fgd7ptruqhv9htgq5jgcfrpv65/1448913600000/01371704680682440553/01371704680682440553/0B-TenPiSMBplUTNya2VoVkI5cGc?e=download&gd=true","editable":true,"etag":"\"fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA\"","fileExtension":"gz","id":"0B-TenPiSMBplUTNya2VoVkI5cGc","labels":{"viewed":true},"lastViewedByMeDate":"2015-11-30T21:37:41.000Z","md5Checksum":"d41d8cd98f00b204e9800998ecf8427e","mimeType":"application/octet-stream","modifiedDate":"2015-11-30T21:37:41.000Z","parents":[{"id":"0B-TenPiSMBplblFKWHQxZGpLaDA","kind":"drive#parentReference","parentLink":"https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplblFKWHQxZGpLaDA","selfLink":"https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc/parents/0B-TenPiSMBplblFKWHQxZGpLaDA"}],"title":"polymap-0.1-2.tar.gz"}
    [699.367376] curl: data out: 0

    [699.745942] curl: header in: HTTP/1.1 200 OK
    [699.746004] curl: header in: X-GUploader-UploadID: AEnB2UpvDEWD6EsNNjY9zFcvZF1WaAm_tTZAVwfMBMoeC0zw8_-D87_Y5-PvIjn26TEk1p6cKXY9zDvKchIPt_MuLJTsHexemw
    [699.746023] curl: header in: Location: https://www.googleapis.com/resumable/upload/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc?fields=alternateLink%2CcreatedDate%2CdownloadUrl%2Ceditable%2Cetag%2CexplicitlyTrashed%2CexportLinks%2CfileExtension%2CfileSize%2Cid%2Clabels%2ClastViewedByMeDate%2Cmd5Checksum%2CmimeType%2CmodifiedDate%2Cparents%2Ctitle&updateViewedDate=true&upload_id=AEnB2UpvDEWD6EsNNjY9zFcvZF1WaAm_tTZAVwfMBMoeC0zw8_-D87_Y5-PvIjn26TEk1p6cKXY9zDvKchIPt_MuLJTsHexemw
    [699.746057] curl: header in: ETag: "fbeGFVkCD3tGFpdjp1CYyuwDABw/mLNL2XBVN-aH8ZKM7yhxWd_PaPg"
    [699.746070] curl: header in: Vary: Origin
    [699.746081] curl: header in: Vary: X-Origin
    [699.746092] curl: header in: Cache-Control: no-cache, no-store, max-age=0, must-revalidate
    [699.746104] curl: header in: Pragma: no-cache
    [699.746114] curl: header in: Expires: Fri, 01 Jan 1990 00:00:00 GMT
    [699.746125] curl: header in: Date: Mon, 30 Nov 2015 21:37:44 GMT
    [699.746137] curl: header in: Content-Length: 0
    [699.746154] curl: info: Server UploadServer is not blacklisted
    [699.746176] curl: header in: Server: UploadServer
    [699.746189] curl: header in: Content-Type: text/html; charset=UTF-8
    [699.746201] curl: header in: Alternate-Protocol: 443:quic,p=1
    [699.746212] curl: header in: Alt-Svc: quic=":443"; ma=604800; v="30,29,28,27,26,25"
    [699.746224] curl: header in: 
    [699.746244] curl: info: Connection #28 to host www.googleapis.com left intact
    [699.747625] curl: info: Found bundle for host www.googleapis.com: 0x16bc690
    [699.747679] curl: info: Re-using existing connection! (#28) with host www.googleapis.com
    [699.747697] curl: info: Connected to www.googleapis.com (216.58.209.106) port 443 (#28)
    [699.747938] curl: header out: PUT /resumable/upload/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc?fields=alternateLink%2CcreatedDate%2CdownloadUrl%2Ceditable%2Cetag%2CexplicitlyTrashed%2CexportLinks%2CfileExtension%2CfileSize%2Cid%2Clabels%2ClastViewedByMeDate%2Cmd5Checksum%2CmimeType%2CmodifiedDate%2Cparents%2Ctitle&updateViewedDate=true&upload_id=AEnB2UpvDEWD6EsNNjY9zFcvZF1WaAm_tTZAVwfMBMoeC0zw8_-D87_Y5-PvIjn26TEk1p6cKXY9zDvKchIPt_MuLJTsHexemw HTTP/1.1
    User-Agent: google-drive-ocamlfuse (0.5.19) gapi-ocaml/0.2.6/Unix
    Host: www.googleapis.com
    Accept: */*
    Accept-Encoding: identity
    Transfer-Encoding: chunked
    Authorization: Bearer 
    If-Match: "fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2MTYzNA"
    Content-Type: application/octet-stream
    Content-Range: bytes 0-1996/1997
    Expect: 100-continue

    [699.804870] curl: header in: HTTP/1.1 100 Continue
    [699.805190] curl: data out: 7cd
    ...
    [700.545312] curl: header in: HTTP/1.1 200 OK
    [700.545380] curl: header in: X-GUploader-UploadID: AEnB2UpvDEWD6EsNNjY9zFcvZF1WaAm_tTZAVwfMBMoeC0zw8_-D87_Y5-PvIjn26TEk1p6cKXY9zDvKchIPt_MuLJTsHexemw
    [700.545394] curl: header in: ETag: "fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2NTA3OA"
    [700.545405] curl: header in: Vary: Origin
    [700.545414] curl: header in: Vary: X-Origin
    [700.545426] curl: header in: Content-Type: application/json; charset=UTF-8
    [700.545435] curl: header in: Cache-Control: no-cache, no-store, max-age=0, must-revalidate
    [700.545449] curl: header in: Pragma: no-cache
    [700.545539] curl: header in: Expires: Fri, 01 Jan 1990 00:00:00 GMT
    [700.545557] curl: header in: Date: Mon, 30 Nov 2015 21:37:45 GMT
    [700.545568] curl: header in: Content-Length: 1262
    [700.545585] curl: info: Server UploadServer is not blacklisted
    [700.545598] curl: header in: Server: UploadServer
    [700.545608] curl: header in: Alternate-Protocol: 443:quic,p=1
    [700.545619] curl: header in: Alt-Svc: quic=":443"; ma=604800; v="30,29,28,27,26,25"
    [700.545643] curl: header in: 
    [700.545657] curl: data in: {
     "id": "0B-TenPiSMBplUTNya2VoVkI5cGc",
     "etag": "\"fbeGFVkCD3tGFpdjp1CYyuwDABw/MTQ0ODkxOTQ2NTA3OA\"",
     "alternateLink": "https://drive.google.com/file/d/0B-TenPiSMBplUTNya2VoVkI5cGc/view?usp=drivesdk",
     "title": "polymap-0.1-2.tar.gz",
     "mimeType": "application/x-gzip",
     "labels": {
      "starred": false,
      "hidden": false,
      "trashed": false,
      "restricted": false,
      "viewed": true
     },
     "createdDate": "2015-11-30T21:37:41.634Z",
     "modifiedDate": "2015-11-30T21:37:45.078Z",
     "lastViewedByMeDate": "2015-11-30T21:37:41.000Z",
     "parents": [
      {
       "kind": "drive#parentReference",
       "id": "0B-TenPiSMBplblFKWHQxZGpLaDA",
       "selfLink": "https://www.googleapis.com/drive/v2/files/0B-TenPiSMBplUTNya2VoVkI5cGc/parents/0B-TenPiSMBplblFKWHQxZGpLaDA",
       "parentLink": "https
    [700.545756] curl: data in: ://www.googleapis.com/drive/v2/files/0B-TenPiSMBplblFKWHQxZGpLaDA",
       "isRoot": false
      }
     ],
     "downloadUrl": "https://doc-0k-a0-docs.googleusercontent.com/docs/securesc/2jstq75ksd6djnng7gd4lrkuiaklkreg/ha6129fgd7ptruqhv9htgq5jgcfrpv65/1448913600000/01371704680682440553/01371704680682440553/0B-TenPiSMBplUTNya2VoVkI5cGc?e=download&gd=true",
     "fileExtension": "gz",
     "md5Checksum": "f4c0dbff2cc518f8123b56ca562cf38a",
     "fileSize": "1997",
     "editable": true,
     "explicitlyTrashed": false
    }


As you can see POST sent application/octet-stream (confirmed in the response and in the subsequent GET). But after uploading with PUT, MIME type was returned application/x-gzip.

Could you send me your logs, where you experience the issue, to my e-mail alessandro(dot)strada(at)gmail(dot)com, please? Feel free to remove any sensitive information. I just need to see the sequence of operations. Thanks!

**Comment by astrada** [comment]
I was thinking: maybe the problem is triggered by the tool used to upload the file. I used `cp` but maybe you used something else: `rsync` perhaps?

**Comment by dos1** [comment]
Generally the files were being copied by our Java application, but when we noticed the issue we started debugging it manually with ``cp``.

I figured out that while images and randomly generated data worked well, any tar.gz file was failing.

As I said, it just stopped working at Nov 20 - we haven't changed anything; files copied earlier were there intact, while everything after that was empty.

Given that it just suddenly changed, it's most likely something on Google Drive side and I guess it could be everything - from gradually deployed change on their side, to some strange kind of rate limiting on our account (we're copying quite a lot of tar.gz files there). As adding MIME type worked, I think the real issue is "what google-drive-ocamlfuse does when it gets server error" instead of "what caused the server error", that's why I filled this ticket.

Anyway, sent the log to you, maybe it'll be useful :)
